Malmo/src/TCPClient.cpp

```diff
@@ -92,67 +98,131 @@ namespace malmo
-    
-    std::string SendStringAndGetShortReply(boost::asio::io_service& io_service, const std::string& ip_address, int port, const std::string& message, bool withSizeHeader)
+
+    std::string Rpc::sendStringAndGetShortReply(boost::asio::io_service& io_service, const std::string& ip_address, int port, const std::string& message, bool withSizeHeader)
     {
         const int MAX_PACKET_LENGTH = 1024;
         unsigned char data[MAX_PACKET_LENGTH + 1];
-        
         std::vector<unsigned char> message_vector(message.begin(), message.end());
-        
+
         tcp::resolver resolver(io_service);
         tcp::resolver::query query(ip_address, boost::lexical_cast<std::string>(port));
-        boost::system::error_code ec;
-        tcp::resolver::iterator endpoint_iterator = resolver.resolve(query, ec);
+        tcp::socket socket(io_service);
 
-        if (ec)
-            LOGERROR(LT("Failed to resolve "), ip_address, LT(":"), port, LT(" - "), ec.message());
+        // Set up a deadline timer to close the socket on timeout, aborting any async io.
 
-        tcp::socket socket(io_service);
-        boost::asio::connect(socket, endpoint_iterator, ec);
-        if (ec)
-            LOGERROR(LT("Failed to connect to "), ip_address, LT(":"), port, LT(" - "), ec.message());
+        boost::asio::deadline_timer deadline(io_service, timeout);
+
+        deadline.async_wait([&](const boost::system::error_code& ec) {
+            if (!ec)
+            {
+                LOGERROR(LT("Request/Reply communication timeout."));
+                boost::system::error_code ignored_ec;
+                socket.close(ignored_ec);
+            }
+        });
 
+        // Attempt the resolve & connect. Relying on background worker thread to progress the async io.
+
+        error_code_sync.init_error_code();
+
+        resolver.async_resolve(query, [&](const boost::system::error_code& ec, tcp::resolver::iterator endpoint_iterator) {
+            if (ec)
+            {
+                LOGERROR(LT("Failed to resolve "), ip_address, LT(":"), port, LT(" - "), ec.message());
+                error_code_sync.signal_error_code(ec);
+            }
+            else
+            {
+                if (endpoint_iterator == tcp::resolver::iterator()) {
+                    error_code_sync.signal_error_code(boost::asio::error::fault);
+                }
+                else {
+                    socket.async_connect(endpoint_iterator->endpoint(), boost::bind(&ErrorCodeSync::signal_error_code, &this->error_code_sync, boost::asio::placeholders::error));
+                }
+            }
+        });
+
+        auto error_code = error_code_sync.await_error_code();
+
+        if (error_code || !socket.is_open())
+        {
+            LOGERROR(LT("Failed to connect to "), ip_address, LT(":"), port, LT(" - "), error_code.message());
+            throw std::runtime_error("Could not connect  " + (error_code ? error_code : boost::asio::error::operation_aborted).message());
```

Comment:
Am wondering about these runtime_errors we're throwing... I have a feeling that in some cases they will propagate all the way up to the agent code, and kill the agent. (Can't remember off-hand which AgentHost methods end up calling `sendStringAndGetShortReply`, so maybe it's not a problem - eg, the sample python code always wraps startMission in a try/except block, because we expect runtime errors from that, but we never put exception handling around sendCommand. There's not much consistency in our code as it stands, and obviously no documentation about what methods are likely to throw... but one thing at a time!) Some of the socket errors may well be unrecoverable-from... but is there a danger that we'll end up killing the agent due to some occasional socket glitch that could otherwise be logged and ignored? Do you have a feel for how often the throws might happen / what could cause them? At the very least, even if we keep throwing, we should make sure each throw is preceded by a LOG call. (Though I think if the throw happens on a background thread, the logger might not be shut down cleanly, in which case the last log messages are likely to be lost anyway...)

Reply:
My main concern is not blocking either the Agent or the Mincraft client (Mod). There could be repeated failures and I've also noticed that logging these can be unreliable.

Reply:
Fair enough - as long as the stress tests are passing then I'm not too worried!